contract/propertyMarketplace.sol

```diff
@@ -84,192 +135,170 @@ contract PropertyMarketplace {
+    function canModifyProperty(uint _index)
+        public
+        view
+        exist(_index)
+        returns (bool)
     {
-        return (
-            msg.sender==properties[_index].owner ||
-            properties[_index].stockData.sold==0 ||
-            properties[_index].status!=Status.SaleCancelled 
-        );
+        return (msg.sender == properties[_index].owner &&
+            properties[_index].stockData.sold == 0 &&
+            properties[_index].status != Status.SaleCancelled);
     }
 
-
+    /// @dev updates the property price
     function updatePropertyPrice(uint _index, uint _price)
-    public
+        public
+        exist(_index)
+        onlyPropertyOwner(_index)
+        onlySharesNotSold(_index)
+        isNotCancelled(_index)
     {
-        require(msg.sender==properties[_index].owner, "Only the property owner can cancel the sale");
-        require(properties[_index].stockData.sold==0, "The sale cannot be updated if some tokens are already issued");
-        require(properties[_index].status!=Status.SaleCancelled, "The property sale is cancelled");
+        require(_price > 0);
         properties[_index].stockData.price = _price;
     }
 
+    /// @dev cancels the sale of a property
     function cancelPropertySale(uint _index)
-    public
+        public
+        exist(_index)
+        onlyPropertyOwner(_index)
+        onlySharesNotSold(_index)
+        isNotCancelled(_index)
     {
-        require(msg.sender==properties[_index].owner, "Only the property owner can cancel the sale");
-        require(properties[_index].stockData.sold==0, "The sale cannot be cancelled if some tokens are already issued");
         properties[_index].status = Status.SaleCancelled;
     }
 
+    /// @dev updates teh number of shares allocated
     function updatePropertyShares(uint _index, uint _shares)
-    public
+        external
+        exist(_index)
+        onlyPropertyOwner(_index)
+        onlySharesNotSold(_index)
+        isNotCancelled(_index)
     {
-        require(msg.sender==properties[_index].owner, "Only the property owner can cancel the sale");
-        require(properties[_index].stockData.sold==0, "The sale cannot be updated if some tokens are already issued");
-        require(properties[_index].status!=Status.SaleCancelled, "The property sale is cancelled");
+        require(_shares > 0, "Invalid number of shares");
         properties[_index].stockData.numShares = _shares;
     }
-    
-    function buyProperty(uint _index) public payable  {
-         require(
-            properties[_index].status!=Status.SaleCancelled
-          ,
-          "This property sale is cancelled."
-        );
 
+    /// @dev buys a share from a property
+    function buyPropertyShare(uint _index)
+        external
+        payable
+        exist(_index)
+        isNotCancelled(_index)
+    {
+        require(
+            properties[_index].owner != msg.sender,
+            "You can't buy shares off your own property"
```

Comment:
haha i don't mind if they do buy shares of their own property, they can be a part owner with other folks. that's ok